doc/programs/mp4creator.1
.TH "mp4creator" "1" "0.9" "Cisco Systems Inc." "MPEG4IP"
.SH "NAME"
.LP 
mp4creator \- creates and modifies mp4 files
.SH "SYNTAX"
.LP 
mp4creator [\fIoptions\fP] <\fIfilename.mp4\fP>

.SH "DESCRIPTION"
.LP 
This program creates or modifies mp4 files. Supported operations are to add media tracks, add hint tracks, delete a track, and optimize the layout of the mp4 file. Currently the program understands MP3, AAC (ADTS), and MPEG\-4 video (Divx) files. These are known as elementary stream (ES) files.
.LP 
Note MP3 files frequently have embedded tags that list information such as title and artist. These tags are not carried forward into the mp4 file.
.LP 
To add a media track to an mp4 file, use the "\-create=<file>" option to specify the ES file. Note if two files names are given on the command line
.LP 
To add a hint track to an mp4 file you need to know the track id of the media track you want to hint. This can be determined by invoking mp4creator with the "\-list" option to obtain a list of existing tracks and their type. Once you've determined the track id, use the "\-hint=<track\-id>" option to create the hint track. Currently the program understands how to create hint tracks for MPEG audio, and MPEG\-4 video. 
.LP 
When adding tracks, mp4creator does not attempt to optimize the file layout for streaming and playback. This can be done once all the desired tracks have been added with the "\-optimize" option.
.LP 
To delete an undesired track from an existing mp4 file, use the "\-delete=<track\-id>" option.
.LP 
Detailed diagnostics are available via the "\-verbosity=<level>" option. Levels 1\-5 are available.
.SH "OPTIONS"
.LP 
.TP 
\fB\-c, \-create\fR=<\fIfile\fP>
Create a media track using <\fIfile\fP>. File must have either .mp3 .aac .divx or .mp4v extension.
.TP 
\fB\-d, \-delete\fR=<\fItrack\-id\fP>
Delete the specified track and all it's associated data.
.TP 
\fB\-H, \-hint\fR=<\fItrack\-id\fP>
Create a hint track for the specified media track. Note this option can be used when creating the media track in which case no track id is necessary. The appropriate RTP payload format is chosen based on the media track content.
.TP 
\fB\-I, \-interleave\fR
Use an interleaved RTP payload format for an audio hint track. Currently this is only valid for AAC audio hint tracks.
.TP 
\fB\-l, \-list\fR
Print a list of tracks in the specified mp4 file and exit.
.TP 
\fB\-m, \-mtu\fR=<\fIsize\fP>
Specifies the network maximum transmission unit (MTU) to use when creating hint tracks. The default value is 1500.
.TP 
\fB\-O, \-optimize\fR
Optimize the layout of the mp4 file to interleave the data in time, hence reducing disk seeks during streaming or playback. This also arranges the mp4 file so that the media control information is at the beginning of the file. This allows HTTP streaming of the mp4 file.
.TP 
\fB\-r, \-rate\fR=<\fIfps\fP>
Specifies video frame rate (frames per second) when creating a video media track. The value can be either an integer or floating point number. It may be necessary to specify this parameter for some MPEG\-4 video streams due to encoding bugs.
.TP 
\fB\-v, \-verbosity\fR=<\fIlevel\fP>
Causes increasing levels of diagnostics to be printed. A level includes details of all lower levels. Levels are:
.br 
	1	Details concerning reading and writing
.br 
	2	Details about control tables
.br 
	3	Details about media samples
.br 
	4	Details about media hints
.br 
	5	Details about everything
.TP 
\fB\-help\fR
.br 
Prints help information.

.SH "EXAMPLES"
.LP 
To create an audio only mp4 file:
.br 
	mp4creator foo.mp3 foo.mp4
.LP 
To create a video only mp4 file:
.br 
	mp4creator foo.divx foo.mp4
.LP 
To create an audio and video file:
.br 
	mp4creator foo.mp3 foo.mp4
.br 
	mp4creator foo.divx foo.mp4
.LP 
To create an audio and video file, hinted, and optimized for streaming:
.br 
	mp4creator foo.mp3 \-H foo.mp4
.br 
	mp4creator foo.divx \-H \-O foo.mp4
.SH "AUTHORS"
.LP 
David Mackie <dmackie@cisco.com>
.br 
Bill May <wmay@cisco.com>
.SH "SEE ALSO"
.LP 
mp4encode(1) gmp4player(1)


NEWS
12/17/01 Version 0.9.2
    Added VC 6.0 projects for windows
    Upgraded to SDL version 1.2.3
    Added shortcut keys for mp4player video window
    Enhanced player debug reporting
    Added RTP debugging messages in player format.
    Disable player sending RTCP when b=RR:0 is set in SDP
	Added man pages for mp4v2 library API
	Added hint track reading utlities to mp4v2 library
	Fixed bugs in SDP version of IOD
    Enhanced player decode performance for all but ISO codec.

11/7/01 Version 0.9.1
	Command line encoding tools now build and run on Windows
	Added a script, mp4v1regenerate, that can convert most old mp4 files 
		created with original mp4 library to corrected mp4 file format 
		generated by new mp4 library
	Fixed bad casts in new mp4 library when creating hint tracks
	Added man pages for primary programs: 
		gmp4player, mp4creator, mp4encode, mp4live 

11/6/01 Version 0.9
	New implementation of mp4 file library and utlities.
	New implementation of mp4 file creator/hinter.
    Added RTP over RTSP (TCP) mode to player.
    Added player "catch up" code for slower processors
	ISMA compliance fixes
	Fixed RFC 2250 compliance problem
	Updated to DSS 3.0.1
    Update to UCL RTP version 1.2.8

9/17/01 Version 0.8
    Port to FreeBSD, BSD/OS, Mac OS X (including Cocoa UI for Mac OS X)
    Conversion of ISO codec to use automake
    Update to SDL 1.2.2, including BSD/OS CD-ROM enhancements.
    Fix various interoperability bugs.
    Added SDP decode debugging.
    Added HTTP fetch capability of SDP files for player,
        including HTTP redirect.
    Fix player for non-44100Hz AAC streams.
    Added setting of debug level for messages 
        from HTTP, RTSP and SDP libraries from player UI.
    Changed player bytestream to throw ints instead of chars 
        when error occurs.
    Added frame fragmentation to ISMA audio RTP bytestream.
    Added search for MP3 files.
    Added config variables for receive RTP ports for player.
    Added better player support for systems with no audio.
    
8/1/01 Version 0.7
    Added live encoding program, mp4live
    Improved player synchronization of audio and video for OSS 
        supported platforms.
    Improved OpenDivX encoder performance
    Added interleaved audio RTP payload format support
    Fixed a number of bugs with mp4 file library
    Fixed compiling on windows platforms

5/7/01 Version 0.6.2
    Fixes so that package now builds on RedHat 6.1, 6.2, 7.0, and 7.1
    LAME MP3 encoder is now included
    Fixes to play Philips MP4 files correctly
    Enabled MMX routines for OpenDivX encoder (x86 architectures only)
    Extended Darwin PlaylistBroadcaster to control multicast IP TTL.
    Faac now handles UNIX paths correctly
    Fixed several bugs in mp4encode script

4/23/01 Version 0.6.1
    Fixed interoperability between OpenDivX and ISO MPEG-4 video codecs
    Fixed RTSP interoperability problem with Darwin Streaming Server 2.0
    Updated mp4encode script to simplify the encoding process
    Player now keeps history list
    Enabled some post-processing of video in the player
    Added mp4extract utility
    Fixed avi2raw to handle zero length video frames in AVI file

3/30/01 Version 0.6
    Inclusion of the OpenDivx MPEG-4 video encoder/decoder
    Inclusion of MP3 audio in addition to AAC audio
    Inclusion of the Apple Darwin Quicktime Server
    Support for the MP4 file format
    Support for multicast streaming
    Added fullscreen mode to player
    Windows port of the player

2/5/01 Version 0.5
    Initial public release


server/mp4creator/mp4creator.cpp
/*
 * The contents of this file are subject to the Mozilla Public
 * License Version 1.1 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of
 * the License at http://www.mozilla.org/MPL/
 * 
 * Software distributed under the License is distributed on an "AS
 * IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * rights and limitations under the License.
 * 
 * The Original Code is MPEG4IP.
 * 
 * The Initial Developer of the Original Code is Cisco Systems Inc.
 * Portions created by Cisco Systems Inc. are
 * Copyright (C) Cisco Systems Inc. 2001.  All Rights Reserved.
 * 
 * Contributor(s): 
 *		Dave Mackie		dmackie@cisco.com
 */

#define MP4CREATOR_GLOBALS
#include "mp4creator.h"
#include "mpeg4ip_getopt.h"

// forward declarations
void PrintTrackList(const char* mp4FileName, u_int32_t verbosity);
char* MakeTempMp4FileName();

MP4TrackId* CreateMediaTracks(
	MP4FileHandle mp4File, const char* inputFileName);

MP4TrackId CreateHintTrack(
	MP4FileHandle mp4File, MP4TrackId mediaTrackId,
	const char* payloadName, bool interleave);


// external declarations

// track creators
MP4TrackId* AviCreator(
	MP4FileHandle mp4File, const char* aviFileName);

MP4TrackId AacCreator(
	MP4FileHandle mp4File, FILE* inFile);

MP4TrackId Mp3Creator(
	MP4FileHandle mp4File, FILE* inFile);

MP4TrackId Mp4vCreator(
	MP4FileHandle mp4File, FILE* inFile);

// hinters
void RfcIsmaHinter(
	MP4FileHandle mp4File, MP4TrackId mediaTrackId, MP4TrackId hintTrackId,
	bool interleave);

void Rfc2250Hinter(
	MP4FileHandle mp4File, MP4TrackId mediaTrackId, MP4TrackId hintTrackId);

void Rfc3119Hinter(
	MP4FileHandle mp4File, MP4TrackId mediaTrackId, MP4TrackId hintTrackId,
	bool interleave);

void Rfc3016Hinter(
	MP4FileHandle mp4File, MP4TrackId mediaTrackId, MP4TrackId hintTrackId);


// main routine
int main(int argc, char** argv)
{
	char* usageString = 
		"usage: %s <options> <mp4-file>\n"
		"  Options:\n"
		"  -create=<input-file>    Create track from <input-file>\n"
		"    input files can be of type: .aac .mp3 .divx .mp4v\n"
		"  -delete=<track-id>      Delete a track\n"
		"  -hint[=<track-id>]      Create hint track, also -H\n"
		"  -interleave             Use interleaved audio payload format, also -I\n"
		"  -list                   List tracks in mp4 file\n"
		"  -mtu=<size>             MTU for hint track\n"
		"  -optimize               Optimize mp4 file layout\n"
		"  -rate=<fps>             Video frame rate, e.g. 30 or 29.97\n"
		"  -verbose[=[1-5]]        Enable debug messages\n"
		;

	bool doCreate = false;
	bool doDelete = false;
	bool doHint = false;
	bool doList = false;
	bool doOptimize = false;
	bool doInterleave = false;
	char* mp4FileName = NULL;
	u_int32_t verbosity = MP4_DETAILS_ERROR;
	char* inputFileName = NULL;
	char* payloadName = NULL;
	MP4TrackId hintTrackId = MP4_INVALID_TRACK_ID;
	MP4TrackId deleteTrackId = MP4_INVALID_TRACK_ID;

	// begin processing command line
	ProgName = argv[0];
	VideoFrameRate = 0;		// determine from input file
	MaxPayloadSize = 1460;

	while (true) {
		int c = -1;
		int option_index = 0;
		static struct option long_options[] = {
			{ "create", 1, 0, 'c' },
			{ "delete", 1, 0, 'd' },
			{ "help", 0, 0, '?' },
			{ "hint", 2, 0, 'H' },
			{ "interleave", 0, 0, 'I' },
			{ "list", 0, 0, 'l' },
			{ "mtu", 1, 0, 'm' },
			{ "optimize", 0, 0, 'O' },
			{ "payload", 1, 0, 'p' },
			{ "rate", 1, 0, 'r' },
			{ "verbose", 2, 0, 'v' },
			{ NULL, 0, 0, 0 }
		};

		c = getopt_long_only(argc, argv, "c:d:H::Ilm:o:Op:v::",
			long_options, &option_index);

		if (c == -1)
			break;

		switch (c) {
		case 'c':
			doCreate = true;
			inputFileName = optarg;
			break;
		case 'd':
			if (sscanf(optarg, "%u", &deleteTrackId) != 1) {
				fprintf(stderr, 
					"%s: bad track-id specified: %s\n",
					 ProgName, optarg);
				exit(EXIT_COMMAND_LINE);
			}
			doDelete = true;
			break;
		case 'H':
			doHint = true;
			if (optarg) {
				if (sscanf(optarg, "%u", &hintTrackId) != 1) {
					fprintf(stderr, 
						"%s: bad track-id specified: %s\n",
						 ProgName, optarg);
					exit(EXIT_COMMAND_LINE);
				}
			}
			break;
		case 'I':
			doInterleave = true;
			break;
		case 'l':
			doList = true;
			break;
		case 'm':
			u_int32_t mtu;
			if (sscanf(optarg, "%u", &mtu) != 1 || mtu < 64) {
				fprintf(stderr, 
					"%s: bad mtu specified: %s\n",
					 ProgName, optarg);
				exit(EXIT_COMMAND_LINE);
			}
			MaxPayloadSize = mtu - 40;	// subtract IP, UDP, and RTP hdrs
			break;
		case 'O':
			doOptimize = true;
			break;
		case 'p':
			payloadName = optarg;
			break;
		case 'r':
			if (sscanf(optarg, "%f", &VideoFrameRate) != 1) {
				fprintf(stderr, 
					"%s: bad rate specified: %s\n",
					 ProgName, optarg);
				exit(EXIT_COMMAND_LINE);
			}
			break;
		case 'v':
			verbosity |= (MP4_DETAILS_READ | MP4_DETAILS_WRITE);
			if (optarg) {
				u_int32_t level;
				if (sscanf(optarg, "%u", &level) == 1) {
					if (level >= 2) {
						verbosity |= MP4_DETAILS_TABLE;
					} 
					if (level >= 3) {
						verbosity |= MP4_DETAILS_SAMPLE;
					} 
					if (level >= 4) {
						verbosity |= MP4_DETAILS_HINT;
					} 
					if (level >= 5) {
						verbosity = MP4_DETAILS_ALL;
					} 
				}
			}
			break;
		case '?':
			fprintf(stderr, usageString, ProgName);
			exit(EXIT_SUCCESS);
		default:
			fprintf(stderr, "%s: unknown option specified, ignoring: %c\n", 
				ProgName, c);
		}
	}

	// check that we have at least one non-option argument
	if ((argc - optind) < 1) {
		fprintf(stderr, usageString, ProgName);
		exit(EXIT_COMMAND_LINE);
	}

	// if it appears we have two file names, then assume -c for the first
	if ((argc - optind) == 2 && inputFileName == NULL) {
		doCreate = true;
		inputFileName = argv[optind++];
	}

	// point to the specified mp4 file name
	mp4FileName = argv[optind++];

	// warn about extraneous non-option arguments
	if (optind < argc) {
		fprintf(stderr, "%s: unknown options specified, ignoring: ", ProgName);
		while (optind < argc) {
			fprintf(stderr, "%s ", argv[optind++]);
		}
		fprintf(stderr, "\n");
	}

	// consistency checks
	if (doDelete && (doCreate || doHint)) {
		fprintf(stderr, 
			"%s: delete operation must be done separately\n",
			 ProgName);
		exit(EXIT_COMMAND_LINE);
	}

	// end processing of command line

	if (doList) {
		PrintTrackList(mp4FileName, verbosity);
		exit(EXIT_SUCCESS);
	}

	// test if mp4 file exists
	bool mp4FileExists = (access(mp4FileName, F_OK) == 0);

	MP4FileHandle mp4File;

	if (doCreate || doHint) {
		if (!mp4FileExists) {
			if (doCreate) {
				mp4File = MP4Create(mp4FileName, verbosity);
				if (mp4File) {
					MP4SetTimeScale(mp4File, 90000);
				}
			} else {
				fprintf(stderr,
					"%s: can't hint track in file that doesn't exist\n", 
					ProgName);
				exit(EXIT_CREATE_FILE);
			}
		} else {
			mp4File = MP4Modify(mp4FileName, verbosity);
		}

		if (!mp4File) {
			// mp4 library should have printed a message
			exit(EXIT_CREATE_FILE);
		}

		if (doCreate) {
			MP4TrackId* pTrackIds = CreateMediaTracks(mp4File, inputFileName);

			if (pTrackIds && doHint) {
				while (*pTrackIds != MP4_INVALID_TRACK_ID) {
					CreateHintTrack(mp4File, *pTrackIds, 
						payloadName, doInterleave);
					pTrackIds++;
				}
			}
		} else {
			CreateHintTrack(mp4File, hintTrackId, payloadName, doInterleave);
		} 

		if (doCreate) {
			MP4MakeIsmaCompliant(mp4File);
		}

		MP4Close(mp4File);

	} else if (doDelete) {
		if (!mp4FileExists) {
			fprintf(stderr,
				"%s: can't delete track in file that doesn't exist\n", 
				ProgName);
			exit(EXIT_CREATE_FILE);
		}

		mp4File = MP4Modify(mp4FileName, verbosity);
		if (!mp4File) {
			// mp4 library should have printed a message
			exit(EXIT_CREATE_FILE);
		}

		MP4DeleteTrack(mp4File, deleteTrackId);

		MP4Close(mp4File);

		doOptimize = true;	// to purge unreferenced track data
	}

	if (doOptimize) {
		char* outputFileName = MakeTempMp4FileName();
		ASSERT(outputFileName);

		MP4Optimize(mp4FileName, outputFileName, verbosity);

		int rc;

#ifdef _WIN32
		rc = remove(mp4FileName);
		if (rc == 0) {
			rc = rename(outputFileName, mp4FileName);
		}
#else
		rc = rename(outputFileName, mp4FileName);
#endif
		if (rc != 0) {
			fprintf(stderr, "%s: can't overwrite %s, output is in %s\n",
				ProgName, mp4FileName, outputFileName);
		}
	}

	return(EXIT_SUCCESS);
}

MP4TrackId* CreateMediaTracks(MP4FileHandle mp4File, const char* inputFileName)
{
	FILE* inFile = fopen(inputFileName, "rb");

	if (inFile == NULL) {
		fprintf(stderr, 
			"%s: can't open file %s: %s\n",
			ProgName, inputFileName, strerror(errno));
		exit(EXIT_CREATE_MEDIA);
	}

	const char* extension = strrchr(inputFileName, '.');
	if (extension == NULL) {
		fprintf(stderr, 
			"%s: no file type extension\n", ProgName);
		exit(EXIT_CREATE_MEDIA);
	}

	static MP4TrackId trackIds[2] = {
		MP4_INVALID_TRACK_ID, MP4_INVALID_TRACK_ID
	};
	MP4TrackId* pTrackIds = trackIds;

	if (!strcasecmp(extension, ".avi")) {
		fclose(inFile);
		inFile = NULL;

		pTrackIds = AviCreator(mp4File, inputFileName);

	} else if (!strcasecmp(extension, ".aac")) {
		trackIds[0] = AacCreator(mp4File, inFile);

	} else if (!strcasecmp(extension, ".mp3")
	  || !strcasecmp(extension, ".mp1")
	  || !strcasecmp(extension, ".mp2")) {
		trackIds[0] = Mp3Creator(mp4File, inFile);

	} else if (!strcasecmp(extension, ".divx")
	  || !strcasecmp(extension, ".mp4v")
	  || !strcasecmp(extension, ".cmp")) {
		trackIds[0] = Mp4vCreator(mp4File, inFile);

	} else {
		fprintf(stderr, 
			"%s: unknown file type\n", ProgName);
		exit(EXIT_CREATE_MEDIA);
	}

	if (inFile) {
		fclose(inFile);
	}

	return pTrackIds;
}

MP4TrackId CreateHintTrack(MP4FileHandle mp4File, MP4TrackId mediaTrackId,
	const char* payloadName, bool interleave)
{
	if (MP4GetTrackNumberOfSamples(mp4File, mediaTrackId) == 0) {
		fprintf(stderr, 
			"%s: couldn't create hint track, no media samples\n", ProgName);
		return MP4_INVALID_TRACK_ID;
	}

	// create the hint track
	MP4TrackId hintTrackId = MP4AddHintTrack(mp4File, mediaTrackId);

	if (hintTrackId == MP4_INVALID_TRACK_ID) {
		fprintf(stderr, 
			"%s: couldn't create hint track\n", ProgName);
		exit(EXIT_CREATE_HINT);
	}

	// vector out to specific hinters
	const char* trackType = MP4GetTrackType(mp4File, mediaTrackId);

	if (!strcmp(trackType, MP4_AUDIO_TRACK_TYPE)) {
		u_int8_t audioType = MP4GetTrackAudioType(mp4File, mediaTrackId);

		switch (audioType) {
		case MP4_MPEG4_AUDIO_TYPE:
		case MP4_MPEG2_AAC_MAIN_AUDIO_TYPE:
		case MP4_MPEG2_AAC_LC_AUDIO_TYPE:
		case MP4_MPEG2_AAC_SSR_AUDIO_TYPE:
			RfcIsmaHinter(mp4File, mediaTrackId, hintTrackId, interleave);
			break;
		case MP4_MPEG1_AUDIO_TYPE:
		case MP4_MPEG2_AUDIO_TYPE:
			if (payloadName && !strcasecmp(payloadName, "3119")) {
				Rfc3119Hinter(mp4File, mediaTrackId, hintTrackId, interleave);
			} else {
				Rfc2250Hinter(mp4File, mediaTrackId, hintTrackId);
			}
			break;
		default:
			fprintf(stderr, 
				"%s: can't hint non-MPEG4/non-MP3 audio type\n", ProgName);
			exit(EXIT_CREATE_HINT);
		}
	} else if (!strcmp(trackType, MP4_VIDEO_TRACK_TYPE)) {
		u_int8_t videoType = MP4GetTrackVideoType(mp4File, mediaTrackId);

		if (videoType == MP4_MPEG4_VIDEO_TYPE) {
			Rfc3016Hinter(mp4File, mediaTrackId, hintTrackId);
		} else {
			fprintf(stderr, 
				"%s: can't hint non-MPEG4 video type\n", ProgName);
			exit(EXIT_CREATE_HINT);
		}
	} else {
		fprintf(stderr, 
			"%s: can't hint track type %s\n", ProgName, trackType);
		exit(EXIT_CREATE_HINT);
	}

	return hintTrackId;
}

void PrintTrackList(const char* mp4FileName, u_int32_t verbosity)
{
	MP4FileHandle mp4File = MP4Read(mp4FileName, verbosity);

	if (!mp4File) {
		fprintf(stderr, "%s: couldn't open %s: %s", 
			ProgName, mp4FileName, strerror(errno));
		exit(1);
	}

	u_int32_t numTracks = MP4GetNumberOfTracks(mp4File);

	printf("Id\tType\n");
	for (u_int32_t i = 0; i < numTracks; i++) {
		MP4TrackId trackId = MP4FindTrackId(mp4File, i);
		const char* trackType = MP4GetTrackType(mp4File, trackId);
		printf("%u\t%s\n", trackId, trackType);
	}

	MP4Close(mp4File);
}

// there are so many attempts to get this right
// that for portablity reasons, it's best just to roll our own
char* MakeTempMp4FileName()
{
#ifndef _WIN32
	static char tempFileName[64];
	u_int32_t i;
	for (i = getpid(); i < 0xFFFFFFFF; i++) {
		sprintf(tempFileName, "./tmp%u.mp4", i);
		if (access(tempFileName, F_OK) != 0) {
			break;
		}
	}
	if (i == 0xFFFFFFFF) {
		return NULL;
	}
#else
	static char tempFileName[MAX_PATH + 3];
	GetTempFileName(".", // dir. for temp. files 
					"mp4",                // temp. filename prefix 
					0,                    // create unique name 
					tempFileName);        // buffer for name 
#endif

	return tempFileName;
}
